src/ol/render/webgl/MixedGeometryBatch.js

```diff
@@ -0,0 +1,368 @@
+ * The resulting WebGL buffers are stored on the batches as well.
+ *
+ * An important aspect of geometry batches is that there is no guarantee that render instructions and WebGL buffers
+ * are synchronized, i.e. render instructions can describe a new state while WebGL buffers might not have been written yet.
+ * This is why two world-to-screen transforms are stored on each batch: one for the render instructions and one for
+ * the WebGL buffers.
+ */
+class MixedGeometryBatch {
+  constructor() {
+    /**
+     * @type {PolygonGeometryBatch}
+     */
+    this.polygonBatch = {
+      entries: {},
+      geometriesCount: 0,
+      verticesCount: 0,
+      ringsCount: 0,
+      renderInstructions: new Float32Array(0),
+      verticesBuffer: new WebGLArrayBuffer(ARRAY_BUFFER, DYNAMIC_DRAW),
+      indicesBuffer: new WebGLArrayBuffer(ELEMENT_ARRAY_BUFFER, DYNAMIC_DRAW),
+      renderInstructionsTransform: createTransform(),
+      verticesBufferTransform: createTransform(),
+      invertVerticesBufferTransform: createTransform(),
+    };
+
+    /**
+     * @type {PointGeometryBatch}
+     */
+    this.pointBatch = {
+      entries: {},
+      geometriesCount: 0,
+      renderInstructions: new Float32Array(0),
+      verticesBuffer: new WebGLArrayBuffer(ARRAY_BUFFER, DYNAMIC_DRAW),
+      indicesBuffer: new WebGLArrayBuffer(ELEMENT_ARRAY_BUFFER, DYNAMIC_DRAW),
+      renderInstructionsTransform: createTransform(),
+      verticesBufferTransform: createTransform(),
+      invertVerticesBufferTransform: createTransform(),
+    };
+
+    /**
+     * @type {LineStringGeometryBatch}
+     */
+    this.lineStringBatch = {
+      entries: {},
+      geometriesCount: 0,
+      verticesCount: 0,
+      renderInstructions: new Float32Array(0),
+      verticesBuffer: new WebGLArrayBuffer(ARRAY_BUFFER, DYNAMIC_DRAW),
+      indicesBuffer: new WebGLArrayBuffer(ELEMENT_ARRAY_BUFFER, DYNAMIC_DRAW),
+      renderInstructionsTransform: createTransform(),
+      verticesBufferTransform: createTransform(),
+      invertVerticesBufferTransform: createTransform(),
+    };
+  }
+
+  /**
+   * @param {Array<import("../../Feature").default>} features Array of features to add to the batch
+   */
+  addFeatures(features) {
+    for (let i = 0; i < features.length; i++) {
+      this.addFeature(features[i]);
+    }
+  }
+
+  /**
+   * @param {import("../../Feature").default} feature Feature to add to the batch
+   */
+  addFeature(feature) {
+    const geometry = feature.getGeometry();
+    if (!geometry) {
+      return;
+    }
+    this.addGeometry_(geometry, feature);
+  }
+
+  /**
+   * @param {import("../../Feature").default} feature Feature
+   * @return {GeometryBatchItem} Batch item added (or existing one)
+   * @private
+   */
+  addFeatureEntryInPointBatch_(feature) {
+    const uid = getUid(feature);
+    if (!(uid in this.pointBatch.entries)) {
+      this.pointBatch.entries[uid] = {
+        feature: feature,
+        properties: feature.getProperties(),
```

Comment:
`Feature.getProperties()` (inherited from `Object.getProperties()` is a bit inefficient, because it returns a copy of the object's `valuees_`. Since this is only used to pass to the shader callback as 2nd argument, without the guarantee that the user needs it at all, I'd suggest removing that second argument and let the user call `feature.get('myProperty')` in the callback.

Reply:
I agree it makes sense to keep the attribute callback signature minimal for now.  I pushed a commit that removes the second argument and removes this `properties` object from the `GeometryBatchItem` type.